### PR TITLE
Tests, stories and improvements for base-field components

### PR DESCRIPTION
**Pull request by gnapse**
## Short description

This PR adds tests, storybook stories and smaller fixes and improvements to these components:

- `TextField`
- `PasswordField`
- `SelectField`

These are all components that share in common the same layout provided by the underlying scaffolding of the non-public component `BaseField`. The bulk of the fixes and improvements are in fact done in this base shared component.

This PR also adds jsdoc documentation to the `BaseField` props.

## PR Checklist

-   [x] Added tests for bugs / new features
-   [x] Updated docs (storybooks, readme)
-   [x] Executed `npm run validate` and made sure no errors / warnings were shown
-   [ ] Described changes in `CHANGELOG.md`
-   [ ] Bumped version in `package.json` and `package-lock.json` (`npm --no-git-tag-version version <major|minor|patch>`) [ref](https://docs.npmjs.com/cli/v6/commands/npm-version)
-   [ ] Updated all static build artifacts (`npm run build-all`)

## Demo

https://user-images.githubusercontent.com/15199/129802193-1e159d5a-b6bd-4f48-8bd1-2bd46743a46c.mp4

### CODE REVIEW

**Comment by gnapse** [comment]
ping @Bloomca

File: src/new-components/password-field/password-field.test.tsx [hunk 1]
```diff
@@ -0,0 +1,235 @@
...
+        const togglePasswordButton = screen.getByRole('button', {
+            name: 'Toggle password visibility',
+        })
+
+        expect(passwordField).toHaveAttribute('type', 'password')
+
+        userEvent.click(togglePasswordButton)
+        expect(passwordField).toHaveAttribute('type', 'text')
+
+        userEvent.click(screen.getByRole('button', { name: 'Toggle password visibility' }))
```
Comment (Bloomca) [inline src/new-components/password-field/password-field.test.tsx:137]: I think we can either have a separate check for it or just pass a different value to `togglePasswordLabel` property to make sure it renders correct label.

**Review by Bloomca — APPROVED** [review]
Looks good, left couple of questions.

File: src/new-components/base-field/base-field.tsx [hunk 2]
```diff
@@ -51,39 +86,75 @@ function BaseField({
...
+                    </Text>
+                    {auxiliaryLabel ? (
+                        <Box className={styles.auxiliaryLabel} paddingLeft="small">
+                            {auxiliaryLabel}
+                        </Box>
+                    ) : null}
+                </Box>
+                {children(ariaDescribedBy ? { id, 'aria-describedby': ariaDescribedBy } : { id })}
+            </Box>
+            {hint ? (
```
[the remark this answers is not in the document]
Comment (gnapse) [reply src/new-components/base-field/base-field.tsx:150]: Good observation. Actually, what needs to go away is that hint guard and the `fieldBox` variable. I forgot to do so in a refactor.

File: src/new-components/base-field/base-field.tsx [hunk 3]
```diff
@@ -51,39 +86,75 @@ function BaseField({
...
         </Box>
     )
+
+    if (!hint) {
+        return fieldBox
+    }
+
+    return (
+        <Stack space="small">
+            <Box className={className} maxWidth={maxWidth} hidden={hidden}>
```
[the remark this answers is not in the document]
Comment (gnapse) [reply src/new-components/base-field/base-field.tsx:128]: Yup, that's related to what I answered you in your other comment. That `fieldBox` variable needs to go away.

File: src/new-components/password-field/password-field.test.tsx (hunk 1 above)
Comment (gnapse) [reply src/new-components/password-field/password-field.test.tsx:137]: Very good point.

**Pull request merged**